docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+Custom Attestation Data Formats for Open Enclave V3 Updates
+================
+
+As custom attestation plugins are developed and features are requested, new APIs were added to:
+
+1. Query built-in SGX attesters and verifiers to be registered.
+2. Select a format of the evidence that the registered attesters can generate.
+3. Query evidence parameters to be used as optional inputs when generating evidences.
+
+The following API was affected:
+
+1. Built-in SGX attesters and verifiers
+
+This document describes the added and affected APIs as well as keeping the existing support for version 2 of the API. This document assumes the user is familiar with the SGX attestation and [custom attestation](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md) with plugins.
+
+Motivation
+----------
+
+New APIs were added to help customers who would like to extend Open Enclave's current attestation structure with extra information while keep TEE-specific information within attester and verifier plugins.
+
+1. The application can query built-in SGX attesters and verifiers to be registered. This allows the plugin library to return attesters and verifiers based on configuration.
+
+```C
+   // Old
+   oe_register_attester(oe_sgx_attesterX(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterY(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterZ(), NULL, 0);
+
+   // New
+   oe_sgx_get_attesters(&attesters, &attesters_length);
+   for (n = 0; n < attesters_length; n++)
+      oe_register_attester(&attesters[n], NULL, 0);
+```
+
+2. The attester application can rely on the SDK to select a matching evidence format.
+
+```C
+   // Old
+   const oe_uuid_t attester_formats[3] = {{ATTESTERX_UUID}, {ATTESTERY_UUID}, {ATTESTERZ_UUID}};
+   selected_format = NULL;
+   for (n = 0; n < requested_formats_length && !selected_format; n++)
+   {
+       for (m = 0; m < countof(attester_formats) && !selected_format; m++)
+       {
+           if (requested_formats[n] == attester_formats[m])
+               selected_format = &attester_formats[m];
+       }
+   }
+
+   // New
+   oe_select_attester_format(
+       requested_formats,
+       requested_formats_length,
+       &selected_format);
+```
+
+3. The application gets optional evidence parameters from verifiers and stays TEE-agnostic when generating evidences.
+
+```C
+   // Old
+   uint8_t opt_paramsX[] = { OPT_PARAMS_X };
+   uint8_t opt_paramsY[] = { OPT_PARAMS_Y };
+   if (format == ATTESTERX_UUID)
+   {
+      oe_get_evidence(&format, 0, NULL, 0, opt_paramsX, sizeof(opt_paramsX), &evidence, &evidence_size, NULL, NULL);
+   }
+   else if (format == ATTESTERY_UUID)
+   {
+       oe_get_evidence(&format, 0, NULL, 0, opt_paramsY, sizeof(opt_paramsY), &evidence, &evidence_size, NULL, NULL);
+   }
+
+   // New
+   uint8_t* opt_params;
+   size_t opt_params_size;
+   oe_get_evidence_parameters(&format, &opt_params, &opt_params_size);
```

Comment:
I like the attempt to reduce the amount of special casing a user has to do, my concern here is whether this is broadly generalizable. It clearly works in the SGX local attestation case for target_info, but do we always expect that the verifier is solely responsible for additional parameters needed by `oe_get_evidence()`? If so, that's actually a constraint that we should document specifically in terms of guidance to plugin authors.

Reply:
oe_get_evidence_parameters() enables the scenarios that evidences need inputs from verifiers. Works for SGX local and possible other TEEs that need a nonce from verifier.

Application can also contribute to evidence generation by providing custom-claims when calling oe_get_evidence(). We will have plugin samples of various combination.

---
In reply to: [393418479](https://github.com/openenclave/openenclave/pull/2621#discussion_r393418479) [](ancestors = 393418479)

Reply:
This doesn't really speak to my concern. Looking at the other comments on this method in the review, it's clear that the _intended_ use for the method is unclear. It sounds from your reply that you are confirming that `opt_params` for `oe_get_evidence()` is actually `verifier_params` and should never be used by an attester plugin implementation to take other kinds of extensible parameters. 

Similarly, that can also clarify the naming of the target function to make that easier to understand, for example, `oe_get_verifier_settings_for_evidence_type()` is a much clearer expression of intent and where it might be used.

Reply:
We want it to be flexible enough that not only it could address the SGX local attestation need, we expect creative plug-in can use it for other purpose.